[distributed] add support for rpc timeouts in process group agent

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28392 [distributed] add support for rpc timeouts in process group agent**

Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds a timeout parameter to process group agent (in miliseconds) that can be set by the user in Python, and also keeps track of timeouts by maintaining a map from timeout --> requests as suggested by @pritamdamania87. The map must stay in sync with the existing `futures_` map, so we write/delete to this map only when the futures_ map is written/deleted.

Note that this diff does not implement the actual process of detecting/handling timed out RPCs. The next step will be to implement a background thread that polls this map periodically and marks futures that have timed out. After this we can also add relevant unit tests. There's still an open question about how we should handle futures that complete/return from the caller after the timeout.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

rohan-varma committed:
[distributed] add timers to future and timeout to rpc to support future
timeouts

Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds timers to the future object,
provides functionality to check if a future is timed out, and allows
specification of the timeout when initializing rpc. A future diff will check for these timeouts and mark the future completed with an exception indicating that it has timed out.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP][distributed] add timers to future and timeout to rpc to support future timeouts"

timeouts**

timeouts

Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds timers to the future object,
provides functionality to check if a future is timed out, and allows
specification of the timeout when initializing rpc. A future diff will check for these timeouts and mark the future completed with an exception indicating that it has timed out.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP][distributed] add timers to future and timeout to rpc to support future timeouts"

timeouts**

timeouts

Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds timers to the future object,
provides functionality to check if a future is timed out, and allows
specification of the timeout when initializing rpc. A future diff will check for these timeouts and mark the future completed with an exception indicating that it has timed out.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP][distributed] add timers to future and timeout to rpc to support future timeouts"

timeouts**

timeouts

Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds timers to the future object,
provides functionality to check if a future is timed out, and allows
specification of the timeout when initializing rpc. A future diff will check for these timeouts and mark the future completed with an exception indicating that it has timed out.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma commented on test/rpc_test.py:
```diff
@@ -836,6 +855,12 @@ def test_rref_forward_chain(self):
+    @requires_process_group_agent("PROCESS_GROUP rpc backend specific test, skip")
+    def test_timeout_set(self):
+        rpc_timeout_seconds = rpc.get_rpc_timeout().seconds
+        self.assertEqual(rpc_timeout_seconds, 100)
```
I have the default timeout for all RPCs to be 100 seconds for now, but I'm not sure what a good default is. If it's too small, users could get time outs when in fact the operation was going to complete successfully.

pritamdamania87 replied:
Let's use 10s for now but also provide a way to specify infinite timeout. Some RPC frameworks use timeout=0 for this. We should document this as well.

rohan-varma commented on test/rpc_test.py:
```diff
@@ -339,6 +344,21 @@ def test_async_add(self):
+            timeout_function,
+            args=(3,)
+        )
+        time.sleep(1)
```
this is not ideal to do in a test, but I'm not sure how to test that the timer works other than this.

zhaojuanmao commented:
Could we add timeout outside of future class? e.g, in process group agent. One reason is that we will use IValue::Future soon, which does not have timeout setting.

rohan-varma commented:
> Could we add timeout outside of future class? e.g, in process group agent. One reason is that we will use IValue::Future soon, which does not have timeout setting.

We'd need to store some information indicating the (relative) time the future was started, so that we can check if the future has timed out. Would it be okay to have this as an additional structure in ProcessGroupAgent?

pritamdamania87 commented:
The future itself should not have a timeout. ProcessGroupAgent should have a timeout on a request and mark the appropriate future with an "Operation timed out" exception.

pritamdamania87 commented on test/rpc_test.py:
```diff
@@ -836,6 +855,12 @@ def test_rref_forward_chain(self):
+    @requires_process_group_agent("PROCESS_GROUP rpc backend specific test, skip")
+    def test_timeout_set(self):
+        rpc_timeout_seconds = rpc.get_rpc_timeout().seconds
+        self.assertEqual(rpc_timeout_seconds, 100)
```
nit: make 100 a constant in the rpc_api.py that you can use elsewhere.

pritamdamania87 commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -79,14 +81,24 @@ PyObject* rpc_init(PyObject* /* unused */) {
           py::arg("process_group"),
-          py::arg("num_send_recv_threads") = 4)
+          py::arg("num_send_recv_threads") = 4,
+          py::arg("future_timeout") = std::chrono::seconds(100))
```
nit: use a constant.

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -126,7 +127,8 @@ ProcessGroupAgent::ProcessGroupAgent(
       sendMutexes_(pg_->getSize()),
-      threadPool_(numSendRecvThreads) {
+      threadPool_(numSendRecvThreads),
+      futureTimeout_(futureTimeout) {
```
this should be called rpc timeout and not future timeout

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -172,6 +174,10 @@ const WorkerInfo& ProcessGroupAgent::getWorkerInfo(worker_id_t id) const {
   return allWorkerInfo_[id];
 }
 
+const std::chrono::seconds& ProcessGroupAgent::getRpcTimeout() {
```
We should store millisecond precision in this class and not at seconds precision.

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -38,7 +38,8 @@ class ProcessGroupAgent : public RpcAgent {
       std::string workerName,
       std::shared_ptr<c10d::ProcessGroup> pg,
-      int numSendRecvThreads = 4);
+      int numSendRecvThreads = 4,
```
nit: not related to this PR, but make 4 a const too.

pritamdamania87 commented on torch/distributed/rpc/__init__.py:
```diff
@@ -14,7 +14,7 @@ def init_model_parallel(self_name,
                             self_rank=-1,
                             init_method=None,
-                            num_send_recv_threads=4):
+                            num_send_recv_threads=4, rpc_timeout=100):
```
rpc_timeout should be specified by the user in milliseconds and not seconds. Some users might want to set it to 100ms for example.

pritamdamania87 replied:
Can we enforce that the user facing API only accepts `timedelta` as the timeout parameter? This would make the API much cleaner.

satgera replied:
Also probably rename it to rpc_timeout_ms

pritamdamania87 commented on torch/distributed/rpc/__init__.py:
```diff
@@ -39,6 +39,6 @@ def init_model_parallel(self_name,
             init_method(str): backend specific init arguments.
             num_send_recv_threads(int): Number of threads for send/recv work.
```
add documentation for rpc_timeout here and mention what the default value is.

pietern commented on torch/csrc/distributed/rpc/future_message.h:
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include <torch/csrc/distributed/rpc/message.h>
+#include <chrono>
```
Include stdlib before everything else.

pietern commented on torch/csrc/distributed/rpc/future_message.h:
```diff
@@ -22,16 +24,23 @@ struct TORCH_API FutureMessage final {
   // TODO: make message_ an optional field, and get rid of UNKNOWN message type
   Message message_;
+  // clock to measure time elapsed since the future was created.
+  std::chrono::steady_clock clock_;
```
Not used.

pietern commented on torch/csrc/distributed/rpc/future_message.h:
```diff
@@ -11,6 +12,7 @@ namespace rpc {
 // TODO: consider using ivalue::Future.
 struct TORCH_API FutureMessage final {
  public:
+  FutureMessage();
```
`explicit`

pietern reviewed (commented):
Is the caller always responsible for timing out an operation?

satgera commented on torch/csrc/distributed/rpc/future_message.cpp:
```diff
@@ -56,6 +60,24 @@ void FutureMessage::fireCallbacks() {
+  if (completed_) {
+    return false;
+  }
+  const auto now = std::chrono::high_resolution_clock::now();
```
Please use std::chrono::steady_clock from what I recall high_resolution_clock isn't monotonic and may be more expensive than steady_clock.

rohan-varma committed:
Update on "[distributed] add timeout for RPC futures, and ability to set timeout when initializing rpc"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds timers to the future object,
provides functionality to check if a future is timed out, and allows
specification of the timeout when initializing rpc. In the next diff, we will add a separate thread to ProcessGroupAgent to clean up these futures if they have timed out.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] add timeout for RPC futures, and ability to set timeout when initializing rpc"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds timers to the future object,
provides functionality to check if a future is timed out, and allows
specification of the timeout when initializing rpc. In the next diff, we will add a separate thread to ProcessGroupAgent to clean up these futures if they have timed out.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] add timeout for RPC futures, and ability to set timeout when initializing rpc"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds timers to the future object,
provides functionality to check if a future is timed out, and allows
specification of the timeout when initializing rpc. In the next diff, we will add a separate thread to ProcessGroupAgent to clean up these futures if they have timed out.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] add timeout for RPC futures, and ability to set timeout when initializing rpc"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds timers to the future object,
provides functionality to check if a future is timed out, and allows
specification of the timeout when initializing rpc. In the next diff, we will add a separate thread to ProcessGroupAgent to clean up these futures if they have timed out.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] add timeout for RPC futures, and ability to set timeout when initializing rpc"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds timers to the future object,
provides functionality to check if a future is timed out, and allows
specification of the timeout when initializing rpc. In the next diff, we will add a separate thread to ProcessGroupAgent to clean up these futures if they have timed out.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] add support for rpc timeouts in process group agent"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds timers to the future object,
provides functionality to check if a future is timed out, and allows
specification of the timeout when initializing rpc. In the next diff, we will add a separate thread to ProcessGroupAgent to clean up these futures if they have timed out.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/rpc/api.py:
```diff
@@ -129,6 +131,16 @@ def get_worker_info(worker_name=None):
+    Returns:
+        `datetime.timedelta` instance indicating the RPC timeout.
+    """
+    return _agent._get_rpc_timeout()
```
Open question about whether we should expose this to the user or not, but I think since it is configurable by the user we should have a getter for it as well.

rohan-varma committed:
Update on "[distributed] add support for rpc timeouts in process group agent"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds a timeout parameter to process group agent (in miliseconds) that can be set by the user in Python, and also keeps track of timeouts by maintaining a map from timeout --> requests as suggested by @pritamdamania87. The map must stay in sync with the existing `futures_` map, so we write/delete to this map only when the futures_ map is written/deleted.

Note that this diff does not implement the actual process of detecting/handling timed out RPCs. The next step will be to implement a background thread that polls this map periodically marks futures that have timed out. There's still an open question about how we should handle futures that complete/return from the caller after the timeout.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] add support for rpc timeouts in process group agent"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds a timeout parameter to process group agent (in miliseconds) that can be set by the user in Python, and also keeps track of timeouts by maintaining a map from timeout --> requests as suggested by @pritamdamania87. The map must stay in sync with the existing `futures_` map, so we write/delete to this map only when the futures_ map is written/deleted.

Note that this diff does not implement the actual process of detecting/handling timed out RPCs. The next step will be to implement a background thread that polls this map periodically marks futures that have timed out. There's still an open question about how we should handle futures that complete/return from the caller after the timeout.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] add support for rpc timeouts in process group agent"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds a timeout parameter to process group agent (in miliseconds) that can be set by the user in Python, and also keeps track of timeouts by maintaining a map from timeout --> requests as suggested by @pritamdamania87. The map must stay in sync with the existing `futures_` map, so we write/delete to this map only when the futures_ map is written/deleted.

Note that this diff does not implement the actual process of detecting/handling timed out RPCs. The next step will be to implement a background thread that polls this map periodically marks futures that have timed out. After this we can also add relevant unit tests. There's still an open question about how we should handle futures that complete/return from the caller after the timeout.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -355,16 +373,26 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
+            // look up the corresponding future by its time out and request ID,
+            // and remove it from the timeouts map
+            auto& futuresAtTime = futureTimeouts_[futureStartTime];
+            futuresAtTime.erase(
```
This code assumes that there is the desired future in the `futureTimeouts_` map, but this may not be the case when we implement removing futures that have timed out. Such a code path would get triggered when a future has timed out and been marked as so, but it later completes and we process it here. In cases like this, I think that we should just drop the future, since the user's request would have already been completed with a timed out exception.

rohan-varma committed:
Update on "[distributed] add support for rpc timeouts in process group agent"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds a timeout parameter to process group agent (in miliseconds) that can be set by the user in Python, and also keeps track of timeouts by maintaining a map from timeout --> requests as suggested by @pritamdamania87. The map must stay in sync with the existing `futures_` map, so we write/delete to this map only when the futures_ map is written/deleted.

Note that this diff does not implement the actual process of detecting/handling timed out RPCs. The next step will be to implement a background thread that polls this map periodically and marks futures that have timed out. After this we can also add relevant unit tests. There's still an open question about how we should handle futures that complete/return from the caller after the timeout.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

xush6528 commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -89,10 +91,15 @@ PyObject* rpc_init(PyObject* /* unused */) {
           py::arg("name"),
           py::arg("process_group"),
-          py::arg("num_send_recv_threads") = 4)
+          py::arg("num_send_recv_threads") = kDefaultNumSendRecvThreads,
```
Can we define a struct, `RpcOptionos`? I am afraid this argument list will keeping growing.

rohan-varma replied:
Created https://github.com/pytorch/pytorch/issues/28940 to track this. Will do it in a follow up diff.

pietern replied:
Since they're kwargs, does it matter? Having the defaults be defined inline is actually kinda nice.

xush6528 replied:
I would try to reduce the use of kwargs, because that will be a blackbox for users. I prefer a data structure with clearly available field names.

xush6528 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -274,9 +281,19 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
   auto future = std::make_shared<FutureMessage>();
   if (message.isRequest()) {
+    // millisecond level precision of when request started.
+    auto futureStartTime =
```
Would it be more convenient to lookup later, if they are keyed by "end time"? I am considering RPCs with overwriting timeouts.

rohan-varma replied:
When we add support for detecting and marking timed out RPCs, we'd like them to be in the `map` by start time, so that it is faster to look up the RPC(s) which started earliest. Then if we determine that RPC has not timed out, we can save ourselves the linear search, which is beneficial since we can have a large number of RPCs.

rohan-varma committed:
Update on "[distributed] add support for rpc timeouts in process group agent"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds a timeout parameter to process group agent (in miliseconds) that can be set by the user in Python, and also keeps track of timeouts by maintaining a map from timeout --> requests as suggested by @pritamdamania87. The map must stay in sync with the existing `futures_` map, so we write/delete to this map only when the futures_ map is written/deleted.

Note that this diff does not implement the actual process of detecting/handling timed out RPCs. The next step will be to implement a background thread that polls this map periodically and marks futures that have timed out. After this we can also add relevant unit tests. There's still an open question about how we should handle futures that complete/return from the caller after the timeout.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] add support for rpc timeouts in process group agent"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds a timeout parameter to process group agent (in miliseconds) that can be set by the user in Python, and also keeps track of timeouts by maintaining a map from timeout --> requests as suggested by @pritamdamania87. The map must stay in sync with the existing `futures_` map, so we write/delete to this map only when the futures_ map is written/deleted.

Note that this diff does not implement the actual process of detecting/handling timed out RPCs. The next step will be to implement a background thread that polls this map periodically and marks futures that have timed out. After this we can also add relevant unit tests. There's still an open question about how we should handle futures that complete/return from the caller after the timeout.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

xush6528 reviewed (approved):
I saw comments are resolved. Can do incremental improvements in following PRs.

rohan-varma committed:
Update on "[distributed] add support for rpc timeouts in process group agent"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds a timeout parameter to process group agent (in miliseconds) that can be set by the user in Python, and also keeps track of timeouts by maintaining a map from timeout --> requests as suggested by @pritamdamania87. The map must stay in sync with the existing `futures_` map, so we write/delete to this map only when the futures_ map is written/deleted.

Note that this diff does not implement the actual process of detecting/handling timed out RPCs. The next step will be to implement a background thread that polls this map periodically and marks futures that have timed out. After this we can also add relevant unit tests. There's still an open question about how we should handle futures that complete/return from the caller after the timeout.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] add support for rpc timeouts in process group agent"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds a timeout parameter to process group agent (in miliseconds) that can be set by the user in Python, and also keeps track of timeouts by maintaining a map from timeout --> requests as suggested by @pritamdamania87. The map must stay in sync with the existing `futures_` map, so we write/delete to this map only when the futures_ map is written/deleted.

Note that this diff does not implement the actual process of detecting/handling timed out RPCs. The next step will be to implement a background thread that polls this map periodically and marks futures that have timed out. After this we can also add relevant unit tests. There's still an open question about how we should handle futures that complete/return from the caller after the timeout.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] add support for rpc timeouts in process group agent"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds a timeout parameter to process group agent (in miliseconds) that can be set by the user in Python, and also keeps track of timeouts by maintaining a map from timeout --> requests as suggested by @pritamdamania87. The map must stay in sync with the existing `futures_` map, so we write/delete to this map only when the futures_ map is written/deleted.

Note that this diff does not implement the actual process of detecting/handling timed out RPCs. The next step will be to implement a background thread that polls this map periodically and marks futures that have timed out. After this we can also add relevant unit tests. There's still an open question about how we should handle futures that complete/return from the caller after the timeout.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

pritamdamania87 commented on test/rpc_test.py:
```diff
@@ -919,6 +918,12 @@ def test_remote_same_worker(self):
         )
         self.assertEqual(rref_c.to_here().wait(), torch.ones(n, n) + 4)
 
+    @dist_init(setup_model_parallel=True)
```
Don't need to specify `setup_model_parallel=True` anymore.

pritamdamania87 commented on test/rpc_test.py:
```diff
@@ -919,6 +918,12 @@ def test_remote_same_worker(self):
         self.assertEqual(rref_c.to_here().wait(), torch.ones(n, n) + 4)
 
+    @dist_init(setup_model_parallel=True)
+    @requires_process_group_agent("PROCESS_GROUP rpc backend specific test, skip")
```
Would we enable timeouts for all backends in a separate test?

rohan-varma replied:
Yes, I will do it once all the code uses the same defaults (to be done in the next diff)

xush6528 replied:
Should be able to remove this decorator now

rohan-varma replied:
This is done in https://github.com/pytorch/pytorch/pull/29068/files, which consolidates the other rpc backends

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -12,6 +12,10 @@ namespace torch {
 namespace rpc {
 
+constexpr int kDefaultNumSendRecvThreads = 4;
+constexpr std::chrono::milliseconds kDefaultRpcTimeout =
```
Can we put this in RpcAgent.h since all agents should share the same default timeout?

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -355,16 +372,26 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
+                std::find(futuresAtTime.begin(), futuresAtTime.end(), id));
+            if (futuresAtTime.size() == 0) {
+              // remove the key from futureTimeouts_
+              futureTimeouts_.erase(futureStartTime);
```
We will add the background thread in a separate PR?

rohan-varma replied:
yes, i will do it in follow up diffs.

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -12,6 +12,10 @@ namespace torch {
 
+constexpr int kDefaultNumSendRecvThreads = 4;
+constexpr std::chrono::milliseconds kDefaultRpcTimeout =
+    std::chrono::seconds(10);
```
distributed autograd, rref and other internal RPCs would end up using this timeout as well. Lets make the default 60s to be conservative. Ideally, this timeout should only affect user RPCs and not system RPCs.

We can do this in future PRs:

1. Have a `timeout` parameter in RpcAgent.send().
2. For user RPCs, we always fill that in with the default rpc timeout.
3. For system RPCs, it'll default to 0 (which would be infinite) unless the system RPC sets it.
4. Later we can allow the user level RPC API to specify timeouts as well.

rohan-varma replied:
Created https://github.com/pytorch/pytorch/issues/29018 to track this.

pietern commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -274,9 +281,19 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
+        futureTimeouts_[futureStartTime].emplace_back(requestId);
+      } else {
+        futureTimeouts_[futureStartTime] = std::vector<int64_t>{requestId};
+      }
```
Why the two different paths?

The `operator[]` function will always return a reference. If the value doesn't yet exist it is default constructed. Instead of first finding if an entry with the start time already exists, you can simply do:

```c++
futureTimeouts_[futureStartTime].push_back(requestId);
```

Also no need to use `emplace_back` here because we're dealing with a trivial type.

pietern commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -50,6 +55,9 @@ class ProcessGroupAgent : public RpcAgent {
   void start() override;
 
+  // retrieves the timeout for all RPCs
+  const std::chrono::milliseconds& getRpcTimeout();
```
This is a const function.

pietern commented on torch/distributed/rpc/api.py:
```diff
@@ -65,7 +66,8 @@ def _init_rpc(
     worker_name_to_id=None,
-    num_send_recv_threads=4,
+    num_send_recv_threads=DEFAULT_NUM_SEND_RECV_THREADS,
+    rpc_timeout=DEFAULT_RPC_TIMEOUT,
```
If these defaults are set someplace else (in C++ land), why override them here?

It's risky, because you can easily end up with 2 sets of defaults, and then you never know which is in effect.

rohan-varma replied:
This is true, right now we'd have to manually make sure they stay in sync. One issue is, I'd want to have default arguments for this in python as well, otherwise, the user will always have to pass it in which is cumbersome. I guess with this particular method `_init_rpc` we don't need to override with defaults since it is only called by `_init_model_parallel`, though, so we can remove the defaults here.

xush6528 replied:
I think we can try first put all timeout defaults in Python world. 

C++ constructors always take a known timeout.

rohan-varma committed:
Update on "[distributed] add support for rpc timeouts in process group agent"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds a timeout parameter to process group agent (in miliseconds) that can be set by the user in Python, and also keeps track of timeouts by maintaining a map from timeout --> requests as suggested by @pritamdamania87. The map must stay in sync with the existing `futures_` map, so we write/delete to this map only when the futures_ map is written/deleted.

Note that this diff does not implement the actual process of detecting/handling timed out RPCs. The next step will be to implement a background thread that polls this map periodically and marks futures that have timed out. After this we can also add relevant unit tests. There's still an open question about how we should handle futures that complete/return from the caller after the timeout.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] add support for rpc timeouts in process group agent"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds a timeout parameter to process group agent (in miliseconds) that can be set by the user in Python, and also keeps track of timeouts by maintaining a map from timeout --> requests as suggested by @pritamdamania87. The map must stay in sync with the existing `futures_` map, so we write/delete to this map only when the futures_ map is written/deleted.

Note that this diff does not implement the actual process of detecting/handling timed out RPCs. The next step will be to implement a background thread that polls this map periodically and marks futures that have timed out. After this we can also add relevant unit tests. There's still an open question about how we should handle futures that complete/return from the caller after the timeout.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] add support for rpc timeouts in process group agent"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds a timeout parameter to process group agent (in miliseconds) that can be set by the user in Python, and also keeps track of timeouts by maintaining a map from timeout --> requests as suggested by @pritamdamania87. The map must stay in sync with the existing `futures_` map, so we write/delete to this map only when the futures_ map is written/deleted.

Note that this diff does not implement the actual process of detecting/handling timed out RPCs. The next step will be to implement a background thread that polls this map periodically and marks futures that have timed out. After this we can also add relevant unit tests. There's still an open question about how we should handle futures that complete/return from the caller after the timeout.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

xush6528 reviewed (changes requested):
Will do a final pass at the end

rohan-varma committed:
Update on "[distributed] add support for rpc timeouts in process group agent"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds a timeout parameter to process group agent (in miliseconds) that can be set by the user in Python, and also keeps track of timeouts by maintaining a map from timeout --> requests as suggested by @pritamdamania87. The map must stay in sync with the existing `futures_` map, so we write/delete to this map only when the futures_ map is written/deleted.

Note that this diff does not implement the actual process of detecting/handling timed out RPCs. The next step will be to implement a background thread that polls this map periodically and marks futures that have timed out. After this we can also add relevant unit tests. There's still an open question about how we should handle futures that complete/return from the caller after the timeout.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma commented:
@xush6528 updated to remove defaults from C++. Could you take another look? Thanks!

xush6528 commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -76,7 +76,7 @@ def process_group_init_backend_handler(
             )
         # TODO: add try-except and destroy _agent in all processes if any fails.
-        return dist.ProcessGroupAgent(self_name, group, num_send_recv_threads)
+        return dist.ProcessGroupAgent(self_name, group, num_send_recv_threads, kwargs["rpc_timeout"])
```
Can we put this into the argument list instead, I think this arg is universal to all RPC backends. So shouldn't unpack it from kwargs.

As regard to `num_send_recv_threads`, it's an arg specific to `ProcessGroupAgent`, we could later remove it from the explicit argument list, and allow users to pass it (like any other extra args) through `kwargs`.

rohan-varma replied:
This is done in https://github.com/pytorch/pytorch/pull/29068/files.

xush6528 commented on torch/distributed/rpc/constants.py:
```diff
@@ -0,0 +1,5 @@
+from datetime import timedelta
+
+
+DEFAULT_NUM_SEND_RECV_THREADS = 4
```
I think we should be clear that this is a arg specific to ProcessGroupAgent.

Putting these 2 in parallel is confusing users would assume they can change the number of threads for other RPC backends as well.

rohan-varma replied:
This is done in https://github.com/pytorch/pytorch/pull/29068/

xush6528 reviewed (approved):
I saw we are resolving the remaining small issues in the following PR.

rohan-varma committed:
Update on "[distributed] add support for rpc timeouts in process group agent"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds a timeout parameter to process group agent (in miliseconds) that can be set by the user in Python, and also keeps track of timeouts by maintaining a map from timeout --> requests as suggested by @pritamdamania87. The map must stay in sync with the existing `futures_` map, so we write/delete to this map only when the futures_ map is written/deleted.

Note that this diff does not implement the actual process of detecting/handling timed out RPCs. The next step will be to implement a background thread that polls this map periodically and marks futures that have timed out. After this we can also add relevant unit tests. There's still an open question about how we should handle futures that complete/return from the caller after the timeout.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] add support for rpc timeouts in process group agent"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds a timeout parameter to process group agent (in miliseconds) that can be set by the user in Python, and also keeps track of timeouts by maintaining a map from timeout --> requests as suggested by @pritamdamania87. The map must stay in sync with the existing `futures_` map, so we write/delete to this map only when the futures_ map is written/deleted.

Note that this diff does not implement the actual process of detecting/handling timed out RPCs. The next step will be to implement a background thread that polls this map periodically and marks futures that have timed out. After this we can also add relevant unit tests. There's still an open question about how we should handle futures that complete/return from the caller after the timeout.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] add support for rpc timeouts in process group agent"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds a timeout parameter to process group agent (in miliseconds) that can be set by the user in Python, and also keeps track of timeouts by maintaining a map from timeout --> requests as suggested by @pritamdamania87. The map must stay in sync with the existing `futures_` map, so we write/delete to this map only when the futures_ map is written/deleted.

Note that this diff does not implement the actual process of detecting/handling timed out RPCs. The next step will be to implement a background thread that polls this map periodically and marks futures that have timed out. After this we can also add relevant unit tests. There's still an open question about how we should handle futures that complete/return from the caller after the timeout.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] add support for rpc timeouts in process group agent"


Per #25531, we want to clean up futures when we detect that there are
failures/timeouts. As a first step, this diff adds a timeout parameter to process group agent (in miliseconds) that can be set by the user in Python, and also keeps track of timeouts by maintaining a map from timeout --> requests as suggested by @pritamdamania87. The map must stay in sync with the existing `futures_` map, so we write/delete to this map only when the futures_ map is written/deleted.

Note that this diff does not implement the actual process of detecting/handling timed out RPCs. The next step will be to implement a background thread that polls this map periodically and marks futures that have timed out. After this we can also add relevant unit tests. There's still an open question about how we should handle futures that complete/return from the caller after the timeout.

Differential Revision: [D18025163](https://our.internmc.facebook.com/intern/diff/D18025163/)

[ghstack-poisoned]
